### ISSUE TITLE
Documentation should call out difference between include tag and include() function

### ISSUE THREAD

**Issue by GPHemsley** [twigphp/Twig#2829]
The `include()` function was added in f4f88a51e275d380663b04bd371b215ef70bb7f0 (10 months after #605 was closed as wontfix), with a line in the changelog that says that it has identical functionality to the `include` tag. #1899 suggested that the `include` tag be deprecated in favor of the `include()` function, which was declined, in part because of the suggestion that the output of the `include()` function could be assigned to a variable. However, the response to #1095 suggests that that is not the case.

So at this point, I am unclear whether there is any difference between the `include` tag and the `include()` function, and the documentation does not even indicate that they both exist for generally the same purpose.

At the very least, I think the following pages should be updated to mention each other:

https://twig.symfony.com/doc/2.x/tags/include.html
https://twig.symfony.com/doc/2.x/functions/include.html

**Comment by stof** [comment]
#1095 was a bug that has been fixed. The `include` function properly returns the rendered content since May 2013.

**Comment by fabpot** [comment]
I've added some notes about this topic in #2834. @GPHemsley How does that sound?

**Comment by kleinfreund** [comment]
As far as I can tell, there is a difference between setting a variable to the result of the include tag compared to setting a variable to the result of an include function call.

In a template where I replaced the former style with the latter, the result changed from rendered HTML to escaped HTML.

Am I doing something wrong? If not, the note on the https://twig.symfony.com/doc/3.x/tags/include.html page is misleading as it indicates the behavior is identical.

I work around this by using an include macro

```twig
{% macro _include(path, properties) %}
  {% include path with properties %}
{% endmacro %}
```

**Comment by stof** [comment]
please show your 2 code snippets.

**Comment by kleinfreund** [comment]
Below are four examples comparing different styles of using include tags and functions. My goal is essentially to be able to nest includes and avoiding the use of set blocks like in example 1 and 2. Example 3 gets me there via a macro and I thought I could use include functions (example 4) to avoid even the macro. But it looks like the set blocks (and also the macro) behave differently when it comes to producing HTML.

```twig
{# Example 1: Include tag + set (pass: renders unescaped HTML) #}

{% set image_1 %}
  {% include "image.twig" with { src: "/img/image.webp" } %}
{% endset %}

{% include "figure.twig" with { content: image_1 } %}



{# Example 2: Include function + set (pass: renders unescaped HTML) #}

{% set image_2 %}
  {{ include("image.twig", { src: "/img/image.webp" }) }}
{% endset %}

{{ include("figure.twig", { content: image_2 }) }}



{# Example 3: Include tag nested via macro (pass: renders unescaped HTML) #}

{% macro _include(path, properties) %}
  {% include path with properties %}
{% endmacro %}

{{ _self._include("figure.twig", {
  content: _self._include("image.twig", { src: "/img/image.webp" }),
}) }}



{# Example 4: Include function nested (fail: renders escaped HTML) #}

{{ include("figure.twig", {
  content: include("image.twig", { src: "/img/image.webp" }),
}) }}
```

What figure.twig and image.twig do exactly doesn't matter much. The issue can be seen if both print their include properties in any HTML tags (e.g. `<p>{{ content }}</p>`).

HTML is rendered without being escaped for examples 1 through 3, but in example 4, HTML is escaped first which means I can't use it for my purpose.

---

Note this could well be a user error and be simply "how it works", but with how the docs describe the include function, I expected different results.

**Comment by stof** [comment]
The last case is expected, because the variable passed to the other template is not wrapped in a `TwigMarkup` object. This not not a difference between the tag and the function. This is the difference between using a capturing `{% set %}` block and not using it (of course, you _cannot_ use the tag version in an expression, so it forces you to use the capturing variant of `{% set %}`). The `include` function is marked as safe for auto-escaping, but your case assigns the return value in the context instead. PHP does not have a way to attach metadata on strings to mark them as safe. And boxing all values would have a huge performance impact and would require all filters to be aware of this boxing (which is both a BC break and a much more painful API). That's why this "consider it as safe for auto-escaping" does not automatically impact cases where you assign it in the context and use it later. The capturing `{% set %}` performs boxing with the `Twig\Markup` object.

Note that your example 3 (using a macro in an expression assigning a context value instead of rendering its output) is quite a hack and I'm not sure how reliable it is over time (I'm not sure whether the testsuite covers such usage). I would recommend against using that variant.

**Comment by stof** [comment]
Note that any function or filter marking its output as would suffer from the same issue when assigning its return value to a context variable (try with the `nl2br` filter)

**Comment by kleinfreund** [comment]
Thanks, @stof! That's a very insightful response. 👍🏻 

> Note that your example 3 (using a macro in an expression assigning a context value instead of rendering its output) is quite a hack and I'm not sure how reliable it is over time (I'm not sure whether the testsuite covers such usage). I would recommend against using that variant.

The way I use this macro is with templates that rely solely on the data passed to them via `properties` (the `include` in the macro has `only`). Its only purpose is really just to safe me from writing a large number of set blocks (i.e. example 1's style from https://github.com/twigphp/Twig/issues/2829#issuecomment-2592171663). As far as I could see in the site I'm working on, this behaves identically to using set blocks and I didn't have issues with it so far.

I kind of see how this is somewhat hacky which is in parts why I wanted to use include functions for it.
